File: src/main/java/org/uqbar/examples/microprocesador/scala/parser/ProgramReader.scala
package org.uqbar.examples.microprocesador.scala.parser

import org.uqbar.examples.microprocesador.scala._

class ProgramReader(instructions: Array[OpCode]) extends ByteManipulation {
	var position = 0

	def hasNext = position < instructions.length;

	def readAddress = asAddress(readByte, readByte)

	def readByte = {
		val byte = instructions(position)
		position += 1
		byte
	}
}

File: src/main/java/org/uqbar/examples/microprocesador/scala/decorator/ProgramReader.java
package org.uqbar.examples.microprocesador.scala.decorator;



public class ProgramReader implements InstructionReader  {

	private byte[] instructions;
	private int position;
	
	public ProgramReader(byte[] instructions) {
		this.instructions = instructions;
		this.position = 0;
	}

	@Override
	public boolean hasNext() {
		return position < instructions.length;
	}

	@Override
	public int readAddress() {
		int low = this.readByte();
		int high = this.readByte();
		return (high << 8) + low;
	}

	@Override
	public byte readByte() {
		return this.instructions[this.position++];
	}

	@Override
	public int getPosition() {
		return this.position;
	}

	@Override
	public int currentByte() {
		return this.instructions[this.position];
	}
	
}
